### DISCUSSION TITLE
BUG :: $.breeze.url helper is not working since latest breeze release

### DISCUSSION THREAD

**Discussion by sandipansaha** [breezefront/community#296] [1. Help] [answered]
Hi @vovayatsyuk 

We have noticed that **$.breeze.url** helper is not working since the latest breeze updates.

![image](https://github.com/breezefront/community/assets/4614581/7d4f214a-32b6-4e05-a214-42a00dc55e57)

![image](https://github.com/breezefront/community/assets/4614581/6cbc3835-2790-4b8a-b08c-cfa3a405c3c6)

Also, we have noticed that the it works when the JS merge is enabled, BUT does not work when JS merge is disabled.
Could it be possibly due to `Swissup_Breeze/js/common/url` being loaded dynamically?

![image](https://github.com/breezefront/community/assets/4614581/12617428-f32e-45b5-9f73-d00a580ca16d)

How can we overcome this or something fundamental has been change?
You can check it here - https://master-7rqtwti-vcte4ksdazidy.eu-4.magentosite.cloud/breeze/

Thanks

**Comment by vovayatsyuk** [comment, marked as answer]
Yes, this js was moved to dynamic. You need to update your code similar to Luma:

```js
define(['mage/url'], (url) => {
    // not you can use url.build
    url.build()
    // p.s. $.breeze.url will be available here too
});
```

**Reply by sandipansaha** [reply]
Hi @vovayatsyuk 
That helps. Thanks.

Could you please guide me if I wanted to load the product recommendations JS directly rather dynamically?
The reason is, I have added mixin to both layoutRenderer and abstractRenderer. But with a hard reload the mixin are not loading. But works fine 2nd time load.
What am I doing wrong probably? https://master-7rqtwti-vcte4ksdazidy.eu-4.magentosite.cloud/breeze/

Thanks

**Reply by vovayatsyuk** [reply]
Did you add `mixins` to your xml file?

Since dynamic js was added, you need to add `mixins` property so Breeze will include js files in correct order. Example:

```diff
 <item name="Vendor_Module/js/catalog-add-to-cart" xsi:type="array">
     <item name="path" xsi:type="string">Vendor_Module/path/to/file</item>
+    <item name="mixins" xsi:type="array">
+       <item name="catalogAddToCart" xsi:type="string">catalogAddToCart</item>
+    </item>
 </item>
```

Docs: https://breezefront.com/docs/mixins#example

**Reply by vovayatsyuk** [reply]
I found one more issue. You need to add `import` statement for all components. Otherwise Breeze can't properly load dependencies in correct order. In your case you need define imports for both `Magento_ProductRecommendationsLayout/js/abstractRenderer` and `Swissup_BreezeMagentoEnterprise/js/product-recommendations/layoutRenderer`

Do you use our [integration module](https://github.com/breezefront/module-breeze-magento-enterprise)? I will update it shortly because it's not updated according latest breeze requirements.

**Reply by sandipansaha** [reply]
Hi @vovayatsyuk 
Thank you for the information. 

Yes, I am using the Swissup_BreezeMagentoEnterprise module. I have used the import the way it is defined in Swissup_BreezeMagentoEnterprise.
My XML -

```php
        <referenceBlock name="breeze.js">
            <arguments>
                <argument name="bundles" xsi:type="array">
                    <item name="default" xsi:type="array">
                        <item name="items" xsi:type="array">
                            <item name="Magento_ProductRecommendationsLayout/js/layoutRenderer" xsi:type="array">
                                <item name="import" xsi:type="array">
                                    <item name="Magento_ProductRecommendationsLayout/js/layoutRendererMixin"
                                          xsi:type="string">WilliamsCommerce_MagentoProductRecommendationsExtend/js/breeze/layoutRendererMixin</item>
                                    <item name="abstractRendererMixin"
                                          xsi:type="string">WilliamsCommerce_MagentoProductRecommendationsExtend/js/breeze/abstractRendererMixin</item>
                                </item>
                            </item>
                            <item name="Magento_PageBuilderProductRecommendations/js/pageBuilderRenderer" xsi:type="array">
                                <item name="import" xsi:type="array">
                                    <item name="Magento_ProductRecommendationsLayout/js/pageBuilderRendererMixin"
                                          xsi:type="string">WilliamsCommerce_MagentoProductRecommendationsExtend/js/breeze/pageBuilderRendererMixin</item>
                                    <item name="abstractRendererMixin"
                                          xsi:type="string">WilliamsCommerce_MagentoProductRecommendationsExtend/js/breeze/abstractRendererMixin</item>
                                </item>
                            </item>
                        </item>
                    </item>
                </argument>
            </arguments>
        </referenceBlock>
``` 


Also, I have tried to define the new mixin in the way. In that case it gets the `layoutRenderer` mixin but the `abstractRenderer` mixin does not work. Probably `abstractRenderer` is defined as import but not as individual component or I am doing something wrong?
Please let me know.

![image](https://github.com/breezefront/community/assets/4614581/22a05bd5-e805-4b64-b1d8-1a5cbf606990)

I tried this codes -
```php
        <referenceBlock name="breeze.js">
            <arguments>
                <argument name="bundles" xsi:type="array">
                    <item name="default" xsi:type="array">
                        <item name="items" xsi:type="array">
                            <item name="abstractRendererMixin" xsi:type="array">
                                <item name="path" xsi:type="string">WilliamsCommerce_MagentoProductRecommendationsExtend/js/breeze/abstractRendererMixin</item>
                                <item name="mixins" xsi:type="array">
                                    <item name="abstractRenderer" xsi:type="string">Magento_ProductRecommendationsLayout/js/abstractRenderer</item>
                                </item>
                            </item>
                            <item name="layoutRendererMixin" xsi:type="array">
                                <item name="path" xsi:type="string">WilliamsCommerce_MagentoProductRecommendationsExtend/js/breeze/layoutRendererMixin</item>
                                <item name="mixins" xsi:type="array">
                                    <item name="layoutRenderer" xsi:type="string">Magento_ProductRecommendationsLayout/js/layoutRenderer</item>
                                </item>
                            </item>
                            <item name="pageBuilderRendererMixin" xsi:type="array">
                                <item name="path" xsi:type="string">WilliamsCommerce_MagentoProductRecommendationsExtend/js/breeze/pageBuilderRendererMixin</item>
                                <item name="mixins" xsi:type="array">
                                    <item name="pageBuilderRenderer" xsi:type="string">Magento_PageBuilderProductRecommendations/js/pageBuilderRenderer</item>
                                </item>
                            </item>
                        </item>
                    </item>
                </argument>
            </arguments>
        </referenceBlock>
```

**Reply by vovayatsyuk** [reply]
Just made a [fix in integration module](https://github.com/breezefront/module-breeze-magento-enterprise/commit/a0918bc9786b72ba2f4f0c3a3fc1dba3f4048596). Can you try if it helps?

**Reply by sandipansaha** [reply]
Thanks @vovayatsyuk 

It seems to be working now.
![image](https://github.com/breezefront/community/assets/4614581/e862eac6-0b37-4ba4-9c14-4f4e7da9189e)

However, I believe, you may have forgotten to remove the `abstractRenderer-register` from the `pageBuilderRenderer` section.

![image](https://github.com/breezefront/community/assets/4614581/9745b6e7-25f5-492d-9334-9e7e3905951d)

Its not available in composer yet. Please let me know when it is pushed. Will wait for that.
Thank you again.

**Reply by vovayatsyuk** [reply]
Thanks! Yes, I forgot to remove that one :+1:

**Comment by JDavidVR** [comment]
Hi @vovayatsyuk 

I'm still confused, below is the script I've created for leafleatjs, but not sure how to inject 'define(['mage/url'], (url) => {' ?

FilePath: app/code/LeafletJs/LeafletJs/view/frontend/web/js/script.js
ScriptContent:

```
(function () {
    'use strict';

    $.widget('leafleat', {
        component: 'LeafletJs_LeafletJs/js/leaflet',
        create: function () {
        
        $.ajax({
                url: $.breeze.url.build('map/data/source'),
                method: 'get',
                success: (data, response) => {
                   // marker logic
                },
                complete: (response) => {
                },
                error: (response, error) => {
                    console.log(error);
                },
            });
    });
})();       
```

**Reply by vovayatsyuk** [reply]
In recent breeze versions you can write js just like in Luma:

```js
define(['mage/url'], (url) => {
    'use strict';

    $.widget('leafleat', {
        component: 'LeafletJs_LeafletJs/js/leaflet',
        create: function () {
            $.ajax({
                url: url.build('map/data/source'),
                method: 'get',
                success: (data, response) => {
                   // marker logic
                },
                complete: (response) => {
                },
                error: (response, error) => {
                    console.log(error);
                },
            });
        }
    });
}); 
```
